# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: third_party/rust/rental/tests/simple_mut.rs
# Commit: 16350e08d34f
# Full Hash: 16350e08d34ff06c7f46745c6b160c7a7bda9140
# Author: Dana Keeler <dkeeler@mozilla.com>
# Date: 2019-12-06 04:43:57
# Regressor Bug: 1586855
# File Overlap Count: 6
# Description:
#   bug 1586855 - incorporate CRLite filters into cert_storage r=jcj,kjacobs
#   
#   This patch implements CRLite lookups for TLS server certificate revocation
#   information in telemetry-only mode. It adds a new preference
#   "security.pki.crlite_mode" to control the behavior of this feature. Setting
# ==============================================================================

diff -r 64fd40663930 -r 16350e08d34f third_party/rust/rental/tests/simple_mut.rs
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/third_party/rust/rental/tests/simple_mut.rs	Thu Dec 05 22:41:53 2019 +0000
@@ -0,0 +1,67 @@
+#[macro_use]
+extern crate rental;
+
+
+pub struct Foo {
+	i: i32,
+}
+
+impl Foo {
+	fn try_borrow_mut(&mut self) -> Result<&mut i32, ()> { Ok(&mut self.i) }
+	fn fail_borrow_mut(&mut self) -> Result<&mut i32, ()> { Err(()) }
+}
+
+
+rental! {
+	mod rentals {
+		use super::*;
+
+		#[rental_mut]
+		pub struct SimpleMut {
+			foo: Box<Foo>,
+			iref: &'foo mut i32,
+		}
+	}
+}
+
+
+#[test]
+fn new() {
+	let foo = Foo { i: 5 };
+	let _ = rentals::SimpleMut::new(Box::new(foo), |foo| &mut foo.i);
+
+	let foo = Foo { i: 5 };
+	let sm = rentals::SimpleMut::try_new(Box::new(foo), |foo| foo.try_borrow_mut());
+	assert!(sm.is_ok());
+
+	let foo = Foo { i: 5 };
+	let sm = rentals::SimpleMut::try_new(Box::new(foo), |foo| foo.fail_borrow_mut());
+	assert!(sm.is_err());
+}
+
+
+#[test]
+fn read() {
+	let foo = Foo { i: 5 };
+
+	let sm = rentals::SimpleMut::new(Box::new(foo), |foo| &mut foo.i);
+	let i: i32 = sm.rent(|iref| **iref);
+	assert_eq!(i, 5);
+
+	let iref: &i32 = sm.ref_rent(|iref| *iref);
+	assert_eq!(*iref, 5);
+}
+
+
+#[test]
+fn write() {
+	let foo = Foo { i: 5 };
+
+	let mut sm = rentals::SimpleMut::new(Box::new(foo), |foo| &mut foo.i);
+
+	{
+		let iref: &mut i32 = sm.ref_rent_mut(|iref| *iref);
+		*iref = 12;
+		assert_eq!(*iref, 12);
+	}
+}